src/xxd/xxd.c

```diff
@@ -855,36 +958,122 @@ main(int argc, char *argv[])
       if (hextype == HEX_LITTLEENDIAN)
 	/* last group will be fully used, round up */
 	c = grplen * ((cols + octspergrp - 1) / octspergrp);
       else
 	c = (grplen * cols - 1) / octspergrp;
-      c += addrlen + 3 + p;
-      l[c++] =
-#ifdef __MVS__
-	  (e >= 64)
-#else
-	  (e > 31 && e < 127)
-#endif
-	  ? e : '.';
-      n++;
-      if (++p == cols)
-	{
-	  l[c] = '\n';
-	  l[++c] = '\0';
-	  xxdline(fpo, l, autoskip ? nonzero : 1);
-	  nonzero = 0;
-	  p = 0;
-	}
+
+      if (color)
+        {
+          if (hextype == HEX_BITS)
+            c += addrlen + 3 + p*12;
+          else
+            c = addrlen + 3 + (grplen * cols - 1)/octspergrp + p*12;
+
+          if (hextype == HEX_LITTLEENDIAN)
+            c += 1;
+
+            COLOR_PROLOGUE
+            begin_coloring_char(l,&c,e,ebcdic);
+
+            #ifdef __MVS__
+            if (e >= 64) l[c++] = e;
+            else  l[c++] = '.';
+            #else
+
+            if (ebcdic)
+              e = (e < 64) ? '.' : etoa64[e-64];
+            l[c++] = (e > 31 && e < 127) ? e : '.';
+            #endif
+
+            COLOR_EPILOGUE
+
+        n++;
+        if (++p == cols)
+          {
+            l[c++] = '\n';
+            l[c++] = '\0';
+            xxdline(fpo, l, autoskip ? nonzero : 1);
+            nonzero = 0;
+            p = 0;
+          }
+        }
+      else /*no colors*/
+        {
+          if (ebcdic)
+            e = (e < 64) ? '.' : etoa64[e-64];
+
+          c += addrlen + 3 + p;
+          l[c++] =
+          #ifdef __MVS__
+              (e >= 64)
+          #else
+              (e > 31 && e < 127)
+          #endif
+              ? e : '.';
+          n++;
+          if (++p == cols)
+            {
+              l[c] = '\n';
+              l[++c] = '\0';
+              xxdline(fpo, l, autoskip ? nonzero : 1);
+              nonzero = 0;
+              p = 0;
+            }
+        }
     }
   if (p)
     {
-      l[c] = '\n';
-      l[++c] = '\0';
+      l[c++] = '\n';
+      l[c] = '\0';
+      if (color)
+        {
+        c++;
+
+        int x = p;
+        if (hextype == HEX_LITTLEENDIAN)
+          {
+            int fill = octspergrp - (p % octspergrp);
+            if (fill == octspergrp) fill = 0;
+
+            c = addrlen + 1 + (grplen * (x - (octspergrp-fill))) / octspergrp;
+
+            int i;
+            for (i = 0;i < fill;i++)
+              {
+                COLOR_PROLOGUE
+
+                l[c++] = COLOR_RED;
+                l[c++] = 'm';
+                l[c++] = ' '; /* empty space */
+
+                COLOR_EPILOGUE
+                x++;
+                p++;
+              }
+          }
+
+        if (hextype != HEX_BITS)
+          {
+            c = addrlen + 1 + (grplen * x) / octspergrp;
+            c += cols - p;
+            c += (cols - p) / octspergrp;
+
+            int i;
+            for (i = cols - p;i > 0;i--)
+              {
+                COLOR_PROLOGUE
```

Comment:
You can emit the prologue and epilogue only once and generate the `n` whitespaces.
Also, why is the space printed in red?

Reply:
Color is not important but this is the easiest way to calculate/print the place when using `-e`

Without invisible (red) white spaces:
```
echo -n A | ./xxd -e -R never
00000000:       41                              A
echo -n A | ./xxd -e -R always
00000000:                                        41                              A
```